File: packages/mobile/ios/MediaSync.swift
//
//  MediaSync.swift
//  HomebasePhotos
//
//  Created by Stef Coenen on 19/06/2024.
//

import Foundation
import MMKV
import Photos
import UniformTypeIdentifiers

class MediaSync {
  func syncMedia () async -> Void {
    let isDebug: Bool = true
    let storage = getStorage()

    let syncEnabled: Bool = storage?.bool(forKey: "syncFromCameraRollAsBoolean") ?? false
    let identity: String = storage?.string(forKey: "identity") ?? ""
    let cat: String = storage?.string(forKey: "bx0900") ?? ""
    let sharedSecret: String = storage?.string(forKey: "APSS") ?? ""
    var lastSyncTimeInMiliseconds: Double = storage?.double(forKey: "lastSyncTimeAsNumber") ?? (Date().timeIntervalSince1970 * 1000) - (1000 * 60 * 60 * 24 * 7)
    let forceLowerQuality: Bool = storage?.bool(forKey: "forceLowerQualityAsBoolean") ?? false

    print("[SyncWorker] syncEnabled: " + (syncEnabled ? "True" : "False"));
    print("[SyncWorker] identity: " + identity);
    print("[SyncWorker] CAT: " + cat);
    print("[SyncWorker] SharedSecret: " + sharedSecret);
    print("[SyncWorker] lastSyncTime: " + String(format: "%f", lastSyncTimeInMiliseconds))
    print("[SyncWorker] forceLowerQuality: " + (forceLowerQuality ? "True" : "False"));

    if(!syncEnabled) {
      return
    }

    // Usefull during debug
    // lastSyncTimeInMiliseconds = 946684800000

    var headers = [String: String]()
    headers["bx0900"] = cat

    guard let sharedSecretBytes = Data(base64Encoded: sharedSecret) else {
      print("[SyncWorker] Failed to decode base64 string for sharedSecret")
      return
    }
    let dotYouClient = DotYouClient(apiType: .app, sharedSecret: sharedSecretBytes, identity: identity, headers: headers)

    // Find all photos that have been added since the last sync
    let fetchOptions = PHFetchOptions()
    let lastSyncTimeSeconds = lastSyncTimeInMiliseconds / 1000 - (60 * 30) // 30 minutes buffer
    let maxBatchSize = 100
    fetchOptions.predicate = NSPredicate(format: "creationDate > %@", NSDate(timeIntervalSince1970: lastSyncTimeSeconds))
    fetchOptions.sortDescriptors = [NSSortDescriptor(key: "creationDate", ascending: false)]
    fetchOptions.fetchLimit = maxBatchSize

    let fetchResult = PHAsset.fetchAssets(with: fetchOptions)
    // let fetchResult = PHAsset.fetchAssets(with: .video, options: fetchOptions)

    if isDebug {
      print("[SyncWorker] fetchResult count: \(fetchResult.count)")
    }

    if fetchResult.count == 0 {
      // No new photos to sync so we set current time as last sync time
      storage?.set(Date().timeIntervalSince1970 * 1000, forKey: "lastSyncTimeAsNumber")
    }


    await withTaskGroup(of: Void.self) { taskGroup in
      for index in 0..<fetchResult.count {
        taskGroup.addTask {
          let asset = fetchResult.object(at: index)
          let options = PHContentEditingInputRequestOptions()
          options.isNetworkAccessAllowed = true

          var filePath: String?
          var mimeType: String = ""
          let width: Int = asset.pixelWidth
          let height: Int = asset.pixelHeight
          let timestampInMillis = (asset.creationDate?.timeIntervalSince1970 ?? 0) * 1000
          let identifier = asset.localIdentifier

          if asset.mediaType == .video {
            // Fetch and set video filePath and mimeType

              if let videoURL = await self.getVideoURL(from: asset) {
                filePath = videoURL.path
                mimeType = "video/mp4"
              } else {
                print("[SyncWorker] Error fetching video URL for asset: \(asset.localIdentifier)")
                return;
              }

          } else {
            // Fetch image filePath and mimeType
            let (input, _) = await self.requestContentEditingInput(asset: asset)
            guard let input = input, let fileURL = input.fullSizeImageURL else { return }

            filePath = fileURL.path
            let resources = PHAssetResource.assetResources(for: asset)
            let uniformTypeIdentifier = resources.first?.uniformTypeIdentifier ?? "public.jpeg"

            if #available(iOS 14.0, *) {
              mimeType = UTType(uniformTypeIdentifier)?.preferredMIMEType ?? "image/jpeg"
            } else {
              mimeType = "image/jpeg"
            }
          }

          // Check if we have a valid file path and continue with the upload process
          guard let validFilePath = filePath else {
            print("[SyncWorker] Error: Could not retrieve file path for asset")
            return
          }

          do {
            let result: UploadResult
            if asset.mediaType == .video {
              // Upload video
              result = try await VideoProvider.uploadMedia(dotYouClient: dotYouClient, filePath: validFilePath, timestampInMs: Int64(timestampInMillis), mimeType: mimeType, identifier: identifier, width: width, height: height, forceLowerQuality: true)
            } else {
              // Upload image
              result = try await ImageProvider.uploadMedia(dotYouClient: dotYouClient, filePath: validFilePath, timestampInMs: Int64(timestampInMillis), mimeType: mimeType, identifier: identifier, width: width, height: height, forceLowerQuality: forceLowerQuality)
            }

            // Common error handling and last sync update logic
            if result is SuccessfulUploadResult {
              print("Upload successful!")
              storage?.set(timestampInMillis, forKey: "lastSyncTimeAsNumber")
            } else if result is BadRequestUploadResult, let badResult = result as? BadRequestUploadResult, badResult.errorCode == "existingFileWithUniqueId" {
              print("Upload failed! Media already exists")
              storage?.set(timestampInMillis, forKey: "lastSyncTimeAsNumber")
            } else {
              print("Upload failed! :shrug:")
            }
          } catch {
            print("[SyncWorker] Error uploading media: \(error.localizedDescription)")
          }
        }
        // Wait for the current task to complete before starting the next one
        await taskGroup.waitForAll()
      }


      print("[SyncWorker] MediaSync finished")
      if(fetchResult.count < maxBatchSize) {
        // Everything is processed and the batch was smaller than max, so we set current time as last sync time
        storage?.set(Date().timeIntervalSince1970 * 1000, forKey: "lastSyncTimeAsNumber")
      }
    }
  }

  // https://github.com/ammarahm-ed/react-native-mmkv-storage/pull/343#issuecomment-1908089428
  func getStorage() -> MMKV? {
    let paths = NSSearchPathForDirectoriesInDomains(.libraryDirectory, .userDomainMask, true)
    guard let first = paths.first else { return nil }
    let path = (first as NSString).appendingPathComponent("mmkv")
    MMKV.initialize(rootDir: path as String)
    return MMKV.init(mmapID: "default")
  }

  func requestContentEditingInput(asset: PHAsset) async -> (input: PHContentEditingInput?, info: [AnyHashable : Any]?) {
    await withCheckedContinuation { continuation in
      let options = PHContentEditingInputRequestOptions()
      options.isNetworkAccessAllowed = true
      asset.requestContentEditingInput(with: options) { input, info in
        continuation.resume(returning: (input, info))
      }
    }
  }

  // Function to request AVAsset asynchronously
  func requestVideoAVAsset(_ asset: PHAsset) async -> AVAsset? {
    await withCheckedContinuation { continuation in
      let options = PHVideoRequestOptions()
      options.isNetworkAccessAllowed = true // Allow download from iCloud if needed

      PHImageManager.default().requestAVAsset(forVideo: asset, options: options) { avAsset, _, _ in
        continuation.resume(returning: avAsset)
      }
    }
  }

  // Function to request AVAsset and get video URL
  func getVideoURL(from asset: PHAsset) async -> URL? {
    if let avAsset = await requestVideoAVAsset(asset) as? AVURLAsset {
      return avAsset.url
    }
    return nil
  }
}

File: packages/mobile/android/app/src/main/java/id/homebase/photos/mediasync/MediaSync.java
package id.homebase.photos.mediasync;

import static id.homebase.lib.core.file.DriveFileUploadProvider.isDebug;

import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.provider.MediaStore;
import android.util.Log;

import com.ammarahmed.mmkv.MMKV;

import java.math.BigDecimal;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

import id.homebase.lib.core.ApiType;
import id.homebase.lib.core.DotYouClient;
import id.homebase.lib.core.crypto.CryptoUtil;
import id.homebase.lib.core.file.types.BadRequestUploadResult;
import id.homebase.lib.core.file.types.SuccessfullUploadResult;
import id.homebase.lib.core.file.types.UploadResult;

public class MediaSync {
    private final Context context;

    public MediaSync(Context context) {
        this.context = context;
    }

    public void syncMedia() {
        Log.v(null, "[SyncWorker] doWork");

        System.loadLibrary("rnmmkv");
        MMKV.initialize(this.context);
        MMKV mmkv = MMKV.mmkvWithID("default");

        assert mmkv != null;
        // Loading key-value pairs from MMKV
        boolean syncEnabled = mmkv.decodeInt("syncFromCameraRollAsBoolean", 0) == 1;
        String identity = mmkv.decodeString("identity", "");
        String CAT = mmkv.decodeString("bx0900", "");
        String sharedSecret = mmkv.decodeString("APSS", "");
        double lastSyncTime = mmkv.decodeDouble("lastSyncTimeAsNumber", new Date().getTime() - 1000 * 60 * 60 * 24 * 7);
        boolean forceLowerQuality = mmkv.decodeInt("forceLowerQualityAsBoolean", 0) == 1;

        assert sharedSecret != null;
        assert identity != null;
        assert CAT != null;

        if (isDebug()) {
            Log.v(null, "[SyncWorker] syncEnabled: " + syncEnabled);
            Log.v(null, "[SyncWorker] identity: " + identity);
            Log.v(null, "[SyncWorker] CAT: " + CAT);
            Log.v(null, "[SyncWorker] sharedSecret: " + sharedSecret);
            Log.v(null, "[SyncWorker] lastSyncTime: " + BigDecimal.valueOf(lastSyncTime).toPlainString());
            //lastSyncTime = 1726131813610L;
        }

        if (!syncEnabled) {
            return;
        }

        Map<String, String> headers = new HashMap<>();
        headers.put("bx0900", CAT);

        DotYouClient dotYouClient = new DotYouClient(ApiType.App, CryptoUtil.base64ToByteArray(sharedSecret), identity, headers);

        double lastSyncTimeSeconds = lastSyncTime / 1000 - (60 * 30); // 30 minutes buffer
        String lastSyncTimeString = String.valueOf(lastSyncTimeSeconds);
        int maxBatchSize = 100;

        // Find all photos that have been added since the last sync
        Uri uri = MediaStore.Files.getContentUri("external");
        String[] projection = {MediaStore.Images.Media.DATA, MediaStore.Images.Media.DATE_ADDED, MediaStore.Images.Media.MIME_TYPE, MediaStore.Images.Media._ID, MediaStore.Images.Media.WIDTH, MediaStore.Images.Media.HEIGHT};
        String selection = MediaStore.Images.Media.DATE_ADDED + " > ?";
        String[] selectionArgs = {lastSyncTimeString};
        String sortOrder = MediaStore.Images.Media.DATE_ADDED + " DESC";
        String limit = " LIMIT " + maxBatchSize;

        Cursor cursor = this.context.getContentResolver().query(
                uri.buildUpon().encodedQuery(limit).build(),
                projection,
                selection,
                selectionArgs,
                sortOrder);

        // For each photo, upload it to the server
        if (cursor != null) {
            if (isDebug()) {
                Log.v(null, "[SyncWorker] cursor: " + cursor.getCount());
            }

            if (cursor.getCount() == 0) {
                // No new photos to sync so we set current time as last sync time
                mmkv.encode("lastSyncTimeAsNumber", new Date().getTime());
            }

            while (cursor.moveToNext()) {
                try {
                    String filePath = cursor.getString(cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATA));
                    String timestampInSeconds = cursor.getString(cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATE_ADDED));
                    // parse timestamp in seconds to double in milliseconds
                    long timestampInMillis = Long.parseLong(timestampInSeconds) * 1000L;

                    String mimeType = cursor.getString(cursor.getColumnIndexOrThrow(MediaStore.Images.Media.MIME_TYPE));
                    String identifier = cursor.getString(cursor.getColumnIndexOrThrow(MediaStore.Images.Media._ID));
                    String width = cursor.getString(cursor.getColumnIndexOrThrow(MediaStore.Images.Media.WIDTH));
                    String height = cursor.getString(cursor.getColumnIndexOrThrow(MediaStore.Images.Media.HEIGHT));

                    if (isDebug()) {
                        Log.v(null, "[SyncWorker] MediaItem filePath: " + filePath);
                    }

                    UploadResult result;
                    if (mimeType.startsWith("video/")) {
                        VideoProvider videoProvider = new VideoProvider(context);
                        result = videoProvider.uploadMedia(dotYouClient, filePath, timestampInMillis, mimeType, identifier, width, height, true);
                    } else {
                        result = ImageProvider.uploadMedia(dotYouClient, filePath, timestampInMillis, mimeType, identifier, width, height, forceLowerQuality);
                    }

                    if (result instanceof SuccessfullUploadResult) {
                        Log.v(null, "[SyncWorker] MediaItem uploaded: " + result.toString());
                        mmkv.encode("lastSyncTimeAsNumber", timestampInMillis); // We update the last sync time to the timestamp of the photo so we can continue where we left of if the task is interrupted
                    } else if (result instanceof BadRequestUploadResult) {
                        if (!Objects.equals(((BadRequestUploadResult) result).getErrorCode(), "existingFileWithUniqueId")) {
                            Log.v(null, "[SyncWorker] MediaItem failed to upload: " + result.toString());
                        } else {
                            Log.v(null, "[SyncWorker] MediaItem was already uploaded: " + result.toString());
                            mmkv.encode("lastSyncTimeAsNumber", timestampInMillis); // We update the last sync time to the timestamp of the photo so we can continue where we left of if the task is interrupted
                        }
                    }
                } catch (Exception e) {
                    // Ignore any errors and continue with the next media item
                    Log.e(null, "[SyncWorker] Error uploading photo: " + e.getMessage());
                }
            }

            if (cursor.getCount() < maxBatchSize) {
                // Everything is processed and the batch was smaller than max, so we set current time as last sync time
                mmkv.encode("lastSyncTimeAsNumber", new Date().getTime());
            }

            cursor.close();
        }


    }
}
